index.js

```diff
@@ -92,54 +152,102 @@ Node.prototype.keys = function () {
+            }
+        }, this);
+    } else if (uri.constructor === String || uri.constructor === Array) {
+        this._uri = parsePattern(uri);
+    }
+    this._str = null;
+    if (params) {
+        this.bind(params);
     }
 }
 
-function parsePattern (pattern) {
-    var bits = normalizePath(pattern);
-    // Re-join {/var} patterns
-    for (var i = 0; i < bits.length - 1; i++) {
-        if (bits[i] === '{' && /}$/.test(bits[i+1])) {
-            bits.splice(i, 2, '{/' + bits[i+1]);
+/**
+ * Binds the provided parameter values to URI's variable components
+ *
+ * @param {Object} params the parameters (and their values) to bind
+ * @return {URI} this URI object
+ */
+URI.prototype.bind = function (params) {
+    if (!params || params.constructor !== Object) {
+        // wrong params format
+        return this;
+    }
+    // look only for parameter keys which match
+    // variables in the URI
+    this._uri.forEach(function (item) {
+        if(item && item.constructor === Object && params[item.name]) {
+            item.pattern = params[item.name];
+            // we have changed a value, so invalidate the string cache
+            this._str = null;
         }
+    }, this);
+    return this;
+};
+
+/**
+ * Builds and returns the full, bounded string path for this URI object
+ *
+ * @return {String} the complete path of this URI object
+ */
+URI.prototype.toString = function () {
+    if (this._str) {
+        // there is a cached version of the URI's string
+        return this._str.concat('');
     }
-    // Parse pattern segments and convert them to objects to be consumed by
-    // Node.set().
-    return bits.map(function(bit) {
-        // Support named but fixed values as
-        // {domain:en.wikipedia.org}
-        var m = /^{([+\/])?([a-zA-Z0-9_]+)(?::([^}]+))?}$/.exec(bit);
-        if (m) {
-            if (m[1]) {
-                throw new Error("Modifiers are not yet implemented!");
+    this._str = '';
+    this._uri.forEach(function (item) {
+        if (item.constructor === Object) {
+            if (item.pattern) {
+                // there is a known value for this variable,
+                // so use it
+                this._str += '/' + encodeURIComponent(item.pattern);
+            } else if (item.modifer) {
+                // we are dealing with a modifier, and there
+                // seems to be no value, so simply ignore the
+                // component
+                this._str += '';
+            } else {
+                // we have a variable component, but no value,
+                // so let's just return the variable name
+                this._str += '/{' + item.name + '}';
             }
-            return {
-                modifier: m[1],
-                name: m[2],
-                pattern: m[3]
-            };
         } else {
-            return bit;
+            this._str += '/' + item;
         }
-    });
+    }, this);
+    return this._str.concat('');
```

Comment:
The .concat() should not be needed, as _str is already a string. (Which is good: In V8 strings built up with += are actually optimized with an internal array, and end up faster than an array + join.)